internal/controllers/certmanager.go

```diff
@@ -398,25 +399,55 @@ func (r *Reconciler) reconcileAgentCertificate(ctx context.Context, cert *certv1
 	return nil
 }
 
+var errCertificateModified error = errors.New("certificate has been modified")
+
 func (r *Reconciler) createOrUpdateCertificate(ctx context.Context, cert *certv1.Certificate, owner metav1.Object) error {
-	certSpec := cert.Spec.DeepCopy()
+	specCopy := cert.Spec.DeepCopy()
 	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, cert, func() error {
 		if owner != nil {
 			if err := controllerutil.SetControllerReference(owner, cert, r.Scheme); err != nil {
 				return err
 			}
 		}
-		// Update Certificate spec
-		cert.Spec = *certSpec
+
+		if !cmp.Equal(cert.Spec.CommonName, specCopy.CommonName) &&
+			!cmp.Equal(cert.Spec.DNSNames, specCopy.DNSNames) &&
+			!cmp.Equal(cert.Spec.Duration, specCopy.Duration) &&
+			!cmp.Equal(cert.Spec.RenewBefore, specCopy.RenewBefore) &&
+			!cmp.Equal(cert.Spec.IPAddresses, specCopy.IPAddresses) &&
+			!cmp.Equal(cert.Spec.URIs, specCopy.URIs) &&
+			!cmp.Equal(cert.Spec.EmailAddresses, specCopy.EmailAddresses) &&
+			!cmp.Equal(cert.Spec.SecretName, specCopy.SecretName) &&
+			!cmp.Equal(cert.Spec.SecretTemplate, specCopy.SecretTemplate) &&
+			!cmp.Equal(cert.Spec.IssuerRef, specCopy.IssuerRef) &&
+			!cmp.Equal(cert.Spec.IsCA, specCopy.IsCA) &&
+			!cmp.Equal(cert.Spec.Keystores, specCopy.Keystores) &&
+			!cmp.Equal(cert.Spec.PrivateKey, specCopy.PrivateKey) &&
+			!cmp.Equal(cert.Spec.EncodeUsagesInRequest, specCopy.EncodeUsagesInRequest) &&
+			!cmp.Equal(cert.Spec.Usages, specCopy.Usages) {
+			return errCertificateModified
+		}
+
 		return nil
 	})
 	if err != nil {
+		if err == errCertificateModified {
+			return r.recreateCertificate(ctx, cert, owner)
+		}
 		return err
 	}
 	r.Log.Info(fmt.Sprintf("Certificate %s", op), "name", cert.Name, "namespace", cert.Namespace)
 	return nil
 }
 
+func (r *Reconciler) recreateCertificate(ctx context.Context, cert *certv1.Certificate, owner metav1.Object) error {
+	err := r.deleteCertificate(ctx, cert)
```

Comment:
Do we also need to delete the certificate's secret?

Reply:
I think that makes sense.